nemo_skills/pipeline/openrlhf/ppo.py

```diff
@@ -0,0 +1,446 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from typing import List
+
+import os
+import nemo_run as run
+import typer
+from datetime import datetime
+from dataclasses import dataclass
+from omegaconf import OmegaConf
+from typing import Optional
+
+from nemo_skills.pipeline.openrlhf import openrlhf_app
+from nemo_skills.pipeline import add_task, check_if_mounted, get_cluster_config, run_exp
+from nemo_skills.pipeline.app import app, typer_unpacker
+from nemo_skills.pipeline.generate import wrap_cmd
+from nemo_skills.utils import setup_logging
+
+LOG = logging.getLogger(__file__)
+
+@dataclass
+class PPOOpenRLHFTask:
+    model: str
+    reward_model: str
+    output_dir: str
+    prompt_data: str
+    num_gpus: int
+    num_nodes: int
+    expname: str
+    disable_wandb: bool
+    wandb_project: str
+    timeout: str
+    extra_arguments: str = ""
+    logging_params: str = ""
+
+
+    def get_ray_server_cmd(self, start_cmd, cluster_config):
+        ray_start_cmd = (
+            "if [ \"${SLURM_PROCID:-0}\" = 0 ]; then "
+            "    echo 'Starting head node' && "
+            "    export RAY_raylet_start_wait_time_s=120 && "
+            "    ray start "
+            "        --head "
+            "        --port=6379 "
+            f"       {self.get_ray_server_ports(cluster_config)} && "
+            f"   {start_cmd} ;"
+            "else "
+            "    echo 'Starting worker node' && "
+            "    export RAY_raylet_start_wait_time_s=120 && "
+            "    echo \"Connecting to head node at $SLURM_MASTER_NODE\" && "
+            "    ray start "
+            "        --block "
+            "        --address=$SLURM_MASTER_NODE:6379 "
+            f"       {self.get_ray_server_ports(cluster_config)} ;"
+            "fi"
+        )
+        return ray_start_cmd
+
+    def get_ray_server_ports(self, cluster_config):
+        ports = (
+            "--node-manager-port=12345 "
+            "--object-manager-port=12346 "
+            "--dashboard-port=8265 "
+            "--dashboard-agent-grpc-port=12347 "
+            "--runtime-env-agent-port=12349 "
+            "--metrics-export-port=12350 "
+            "--min-worker-port=14349 "
+            "--max-worker-port=18349 "
+        )
+        return ports
+
+    def get_ray_launch_cmd(self, cluster_config):
+        cmd = (
+            "ray job submit --address='http://127.0.0.1:8265' -- "
+        )
+        return cmd
+
+
+    def format_reward_critic_args(self, cluster_config):
+        cmd = (
+            f" --reward_pretrain {self.reward_model} "
```

Comment:
I don't have a good sense of how these parameters should be set - are there any docs on that? Intuitively it seems that different models should perhaps be located on different sets of nodes, but not sure